libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1280 @@
+  _LIBCPP_HIDE_FROM_ABI
+  flat_map(sorted_unique_t,
+           const key_container_type& __key_cont,
+           const mapped_container_type& __mapped_cont,
+           const key_compare& __comp,
+           const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_tag{}, __alloc, __key_cont, __mapped_cont, __comp) {}
+
+  _LIBCPP_HIDE_FROM_ABI explicit flat_map(const key_compare& __comp) : __containers_(), __compare_(__comp) {}
+
+  template <class _Allocator>
+    requires __allocator_ctor_constraint<_Allocator>
+  _LIBCPP_HIDE_FROM_ABI flat_map(const key_compare& __comp, const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_empty_tag{}, __alloc, __comp) {}
+
+  template <class _Allocator>
+    requires __allocator_ctor_constraint<_Allocator>
+  _LIBCPP_HIDE_FROM_ABI explicit flat_map(const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_empty_tag{}, __alloc) {}
+
+  template <input_iterator _InputIterator>
+  _LIBCPP_HIDE_FROM_ABI
+  flat_map(_InputIterator __first, _InputIterator __last, const key_compare& __comp = key_compare())
+      : __containers_(), __compare_(__comp) {
+    insert(__first, __last);
+  }
+
+  template <input_iterator _InputIterator, class _Allocator>
+    requires __allocator_ctor_constraint<_Allocator>
+  _LIBCPP_HIDE_FROM_ABI
+  flat_map(_InputIterator __first, _InputIterator __last, const key_compare& __comp, const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_empty_tag{}, __alloc, __comp) {
+    insert(__first, __last);
+  }
+
+  template <input_iterator _InputIterator, class _Allocator>
+    requires __allocator_ctor_constraint<_Allocator>
+  _LIBCPP_HIDE_FROM_ABI flat_map(_InputIterator __first, _InputIterator __last, const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_empty_tag{}, __alloc) {
+    insert(__first, __last);
+  }
+
+  template <_ContainerCompatibleRange<value_type> _Range>
+  _LIBCPP_HIDE_FROM_ABI flat_map(from_range_t __fr, _Range&& __rg)
+      : flat_map(__fr, std::forward<_Range>(__rg), key_compare()) {}
+
+  template <_ContainerCompatibleRange<value_type> _Range, class _Allocator>
+    requires __allocator_ctor_constraint<_Allocator>
+  _LIBCPP_HIDE_FROM_ABI flat_map(from_range_t, _Range&& __rg, const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_empty_tag{}, __alloc) {
+    insert_range(std::forward<_Range>(__rg));
+  }
+
+  template <_ContainerCompatibleRange<value_type> _Range>
+  _LIBCPP_HIDE_FROM_ABI flat_map(from_range_t, _Range&& __rg, const key_compare& __comp) : flat_map(__comp) {
+    insert_range(std::forward<_Range>(__rg));
+  }
+
+  template <_ContainerCompatibleRange<value_type> _Range, class _Allocator>
+  _LIBCPP_HIDE_FROM_ABI flat_map(from_range_t, _Range&& __rg, const key_compare& __comp, const _Allocator& __alloc)
+      : flat_map(__ctor_uses_allocator_empty_tag{}, __alloc, __comp) {
+    insert_range(std::forward<_Range>(__rg));
+  }
+
+  template <input_iterator _InputIterator>
```

Comment:
As you noticed, this should be `template <class _InputIterator>` per the spec. If that's meaningful, maybe we can add tests for that.

Reply:
This paragraph says `InputIterator` in the wording means "**input iterator**"

http://eel.is/c++draft/containers#sequence.reqmts-67
> For every sequence container defined in this Clause and in [[strings]](http://eel.is/c++draft/strings):
[(67.1)](http://eel.is/c++draft/containers#sequence.reqmts-67.1)
If the constructor
template<class InputIterator>
  X(InputIterator first, InputIterator last,
    const allocator_type& alloc = allocator_type());
is called with a type InputIterator that does not qualify as an ***input iterator***, then the constructor shall not participate in overload resolution.

And in this paragraph, it seems to day that "**input iterator**" means the concept `input_iterator`.
http://eel.is/c++draft/iterators#iterator.requirements.general-2

> This document defines six categories of iterators, according to the operations defined on them: [**input iterators**](http://eel.is/c++draft/iterators#def:input_iterators), [output iterators](http://eel.is/c++draft/iterators#def:output_iterators), [forward iterators](http://eel.is/c++draft/iterators#def:forward_iterators), [bidirectional iterators](http://eel.is/c++draft/iterators#def:bidirectional_iterators), [random access iterators](http://eel.is/c++draft/iterators#def:random_access_iterators), and [contiguous iterators](http://eel.is/c++draft/iterators#def:contiguous_iterators), as shown in Table [88](http://eel.is/c++draft/iterators#tab:iterators.relations)[.](http://eel.is/c++draft/iterators#iterator.requirements.general-2.sentence-3)

>The six categories of iterators correspond to the iterator concepts
[(3.1)](http://eel.is/c++draft/iterators#iterator.requirements.general-3.1)
[`input_iterator`](http://eel.is/c++draft/iterators#concept:input_iterator) ([[iterator.concept.input]](http://eel.is/c++draft/iterators#iterator.concept.input)),
[(3.2)](http://eel.is/c++draft/iterators#iterator.requirements.general-3.2)
[output_iterator](http://eel.is/c++draft/iterators#concept:output_iterator) ([[iterator.concept.output]](http://eel.is/c++draft/iterators#iterator.concept.output)),
[(3.3)](http://eel.is/c++draft/iterators#iterator.requirements.general-3.3)
[forward_iterator](http://eel.is/c++draft/iterators#concept:forward_iterator) ([[iterator.concept.forward]](http://eel.is/c++draft/iterators#iterator.concept.forward)),
[(3.4)](http://eel.is/c++draft/iterators#iterator.requirements.general-3.4)
[bidirectional_iterator](http://eel.is/c++draft/iterators#concept:bidirectional_iterator) ([[iterator.concept.bidir]](http://eel.is/c++draft/iterators#iterator.concept.bidir)),
[(3.5)](http://eel.is/c++draft/iterators#iterator.requirements.general-3.5)
[random_access_iterator](http://eel.is/c++draft/iterators#concept:random_access_iterator) ([[iterator.concept.random.access]](http://eel.is/c++draft/iterators#iterator.concept.random.access)), and
[(3.6)](http://eel.is/c++draft/iterators#iterator.requirements.general-3.6)
[contiguous_iterator](http://eel.is/c++draft/iterators#concept:contiguous_iterator) ([[iterator.concept.contiguous]](http://eel.is/c++draft/iterators#iterator.concept.contiguous)),

But later here it also says
http://eel.is/c++draft/iterators#input.iterators-1
> A class or pointer type X meets the requirements of an **input iterator** for the value type T if X meets the Cpp17Iterator ([[iterator.iterators]](http://eel.is/c++draft/iterators#iterator.iterators)) and Cpp17EqualityComparable (Table [28](http://eel.is/c++draft/utility.arg.requirements#tab:cpp17.equalitycomparable)) requirements and the expressions in Table [90](http://eel.is/c++draft/iterators#tab:inputiterator) are valid and have the indicated semantics[.](http://eel.is/c++draft/iterators#input.iterators-1.sentence-1)

so, we should be using `__cpp17_input_iterator`

Reply:
That's extremely confusing indeed. It seems like we've made a choice of how to interpret the Standard in e.g. `std::vector` and `std::deque` where we use `std::__has_input_iterator_category` to check for whether a type "is an input iterator". I would stick to that interpretation in `flat_map` & friends for consistency.

If we then go back and decide that this interpretation of the Standard is wrong (and we should be using e.g. `concept input_iterator` instead), then I would change it consistently for all the containers that are affected.